### PR TITLE
Lecture3

### PR DESCRIPTION
**Pull request by YahaJava**

### CODE REVIEW

File: src/main/java/com/es/phoneshop/model/card/CartItem.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package com.es.phoneshop.model.card;
+
+import com.es.phoneshop.model.product.Product;
+
+public class CartItem
+{
```
Comment (ghost) [inline src/main/java/com/es/phoneshop/model/card/CartItem.java]: Please follow the project codestyle

File: src/main/java/com/es/phoneshop/model/card/HttpSessionCartService.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        if (cart == null) {
+            cart = new Cart();
+            request.getSession().setAttribute(CART_ATRIBUTE, cart);
+        }
+        return cart;
+    }
+
+    @Override
+    public void add(Cart cart, Product product, int quantity) {
+        if (quantity > product.getStock()) {
```
Comment (ghost) [inline src/main/java/com/es/phoneshop/model/card/HttpSessionCartService.java]: Let's say we have a product "testProduct" having a stock value of 20. What will happen if we do the following:
cartService.add(cart, testProduct, 20);
cartService.add(cart, testProduct, 1);
?

File: src/main/java/com/es/phoneshop/model/card/HttpSessionCartService.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            throw new OutOfStockException(product.getStock());
+        }
+        cart.getCartItems().add(new CartItem(product, quantity));
```
Comment (ghost) [inline src/main/java/com/es/phoneshop/model/card/HttpSessionCartService.java]: Please, handle cart items the way we would always have only one cart item per product

File: src/main/java/com/es/phoneshop/model/card/HttpSessionCartService.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        recalculate(cart);
+    }
+
+    private void recalculate(Cart cart) {
+        cart.setTotalQuantity(cart.getCartItems().stream()
+                .mapToInt(item -> item.getQuantity())
```
Comment (ghost) [inline src/main/java/com/es/phoneshop/model/card/HttpSessionCartService.java]: Lambda can be replaced with method reference

File: src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java [hunk 5]
```diff
@@ -1,35 +1,86 @@
...
+import java.text.ParseException;
+import java.util.Locale;
 
 
 public class ProductDetailsPageServlet extends HttpServlet {
 
-    private ProductDao products;
+    private ProductDao productDao;
+    private CartService cartService;
+    private ProductHistory productHistory;
```
Comment (ghost) [inline src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java]: Where is ProductHistory class defined? Also, name it something like "ProductHistoryService"

File: src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java [hunk 6]
```diff
@@ -1,35 +1,86 @@
...
     @Override
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws ServletException, IOException {
-        String code = request.getPathInfo().substring(1);
+        String code = getCode(request);
         try {
-            request.setAttribute("product", products.getProduct(code));
-            request.getRequestDispatcher("/WEB-INF/pages/productDetails.jsp").forward(request, response);
+            Product product = productDao.getProduct(code);
+            showPage(request, response, product);
+            if (productHistory.getRecentProducts().size() > 2 & productHistory.save(product)) {
```
Comment (ghost) [inline src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java]: Use && in boolean operations

Comment (ghost) [reply src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java]: \+ Where is productHistory supposed to take recently viewed products from? They must be specific for each customer

Comment (ghost) [reply src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java]: \+ Maybe it makes sense to move all this logic (checking the history size, deleting from history, etc) to productHistoryService itself

File: src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java [hunk 7]
```diff
@@ -1,35 +1,86 @@
...
+                productHistory.delete();
+            }
+
         } catch (IllegalArgumentException e) {
             request.setAttribute("product", code);
```
Comment (juda-s) [inline src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java]: Please make use of web.xml error pages

File: src/main/java/com/es/phoneshop/model/card/Cart.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
+
+package com.es.phoneshop.model.card;
```
Comment (ghost) [inline src/main/java/com/es/phoneshop/model/card/Cart.java]: Cart, not card

File: src/main/java/com/es/phoneshop/model/card/HttpSessionCartService.java (hunk 2 above)
Comment (ghost) [reply src/main/java/com/es/phoneshop/model/card/HttpSessionCartService.java]: Still relevant

Comment (ghost) [reply src/main/java/com/es/phoneshop/model/card/HttpSessionCartService.java]: Additionally, simplify the logic of this class by using Java 8 optionals

File: src/main/java/com/es/phoneshop/model/card/HttpSessionCartService.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
+package com.es.phoneshop.model.card;
+
+import com.es.phoneshop.model.exceptions.OutOfStockException;
+import com.es.phoneshop.model.product.Product;
+
+import javax.servlet.http.HttpServletRequest;
+import java.math.BigDecimal;
+import java.util.Optional;
+
+public class HttpSessionCartService implements CartService {
```
Comment (ghost) [inline src/main/java/com/es/phoneshop/model/card/HttpSessionCartService.java]: Write tests for this class

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 10]
```diff
@@ -6,26 +6,22 @@
...
+        return ProductDaoHolder.productDao;
     }
 
 
     public List<Product> getAllProducts() {
         return products;
     }
 
-    protected void setProduct(Product product) {
+    public void setProduct(Product product) {
```
Comment (ghost) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Remove this method and write a protected setter method for product list to be used in test class

File: src/main/java/com/es/phoneshop/model/product/ProductHistoryService.java [hunk 11]
```diff
@@ -0,0 +1,47 @@
+package com.es.phoneshop.model.product;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ProductHistoryService {
```
Comment (ghost) [inline src/main/java/com/es/phoneshop/model/product/ProductHistoryService.java:7]: Product history must be stored in session as it must be customer-specific

Comment (ghost) [reply src/main/java/com/es/phoneshop/model/product/ProductHistoryService.java:7]: Additionally, cover this class in tests

File: src/main/java/com/es/phoneshop/web/ProductPriceHistoryServlet.java [hunk 12]
```diff
@@ -11,22 +11,22 @@
 
 public class ProductPriceHistoryServlet extends HttpServlet {
```
Comment (ghost) [inline src/main/java/com/es/phoneshop/web/ProductPriceHistoryServlet.java:12]: Cover this class also in tests

**Pull request merged**